controllers/etcdrestore_controller.go

```diff
@@ -0,0 +1,524 @@
+	expectedCluster := clusterForRestore(restore)
+	expectedCluster.Default()
+	// We know that the cluster will be of size one for the restore. So only consider peer zero.
+	expectedPeerNames := expectedPeerNamesForCluster(expectedCluster)
+	expectedPeers := make([]etcdv1alpha1.EtcdPeer, len(expectedPeerNames))
+	for i, peerName := range expectedPeerNames {
+		expectedPeers[i] = *peerForCluster(expectedCluster, peerName)
+		expectedPeers[i].Default()
+		configurePeerBootstrap(&expectedPeers[i], expectedCluster)
+	}
+	expectedPVCs := make([]corev1.PersistentVolumeClaim, len(expectedPeers))
+	for i, peer := range expectedPeers {
+		expectedPVCs[i] = *pvcForPeer(&peer)
+		markPVC(restore, &expectedPVCs[i])
+	}
+
+	// Step 1. Create the PVCs.
+	for _, expectedPVC := range expectedPVCs {
+		var pvc corev1.PersistentVolumeClaim
+		err := r.Get(ctx, name(&expectedPVC), &pvc)
+
+		if apierrors.IsNotFound(err) {
+			// No PVC. Make it!
+			log.Info("Creating PVC", "pvc-name", name(&expectedPVC))
+			err := r.Create(ctx, &expectedPVC)
+			return ctrl.Result{}, err
+		} else if err != nil {
+			// There was some other, non non-found error. Exit as we can't handle this case.
+			log.Info("Encountered error while finding PVC")
+			return ctrl.Result{}, err
+		} else {
+			log.Info("PVC already exists. continuing", "pvc-name", name(&pvc))
+			// The PVC is there already. Continue.
+		}
+
+		// Check to make sure we're expecting to restore into this PVC
+		if !IsOurPVC(restore, pvc) {
+			// It's unsafe to take any further action. We don't control the PVC.
+			log.Info("PVC is not marked as our PVC. Failing", "pvc-name", name(&pvc))
+			// TODO Report error reason
+			restore.Status.Phase = etcdv1alpha1.EtcdRestorePhaseFailed
+			err := r.Client.Status().Update(ctx, &restore)
+			return ctrl.Result{}, err
+		} else {
+			log.Info("PVC correctly marked as ours", "pvc-name", name(&pvc))
+			// Great, this PVC is marked as *our* restore (and not any random PVC, or one for an existing cluster, or
+			// a conflicting restore).
+		}
+	}
+
+	// So the peer restore is like a distributed fork/join. We need to launch n restore pods, where n is the number of
+	// PVCs. Then wait for all of them to be complete. If any fail we abort.
+
+	log.Info("Searching for restore pods")
+	restorePods := make([]*corev1.Pod, len(expectedPeerNames))
+	// Go find the restore pods for each peer
+	for i, peer := range expectedPeers {
+		log.Info("Searching for restore pod for peer", "peer-name", peer.Name)
+		pod := corev1.Pod{}
+		err := r.Get(ctx, restorePodNamespacedName(peer), &pod)
+		if client.IgnoreNotFound(err) != nil {
+			// This was a non-notfound error. Fail!
+			return ctrl.Result{}, err
+		} else if err == nil {
+			restorePods[i] = &pod
+		}
+	}
+
+	log.Info("Finished querying for restore pods")
+	// Verify each pod we found is actually ours
+	for _, restorePod := range restorePods {
+		if restorePod != nil && !IsOurPod(restore, *restorePod) {
+			// TODO Report error reason
```

Comment:
Yeah, this seems important. Maybe create a followup issue and link to it.

Reply:
Ah I meant to tidy this up. I should issue an event in this case.

Reply:
I generate events for everything now.